doc/api/esm.md

```diff
@@ -106,11 +106,20 @@ fs.readFile('./foo.txt', (err, body) => {
 <!-- type=misc -->
 
 To customize the default module resolution, loader hooks can optionally be
-provided via a `--loader ./loader-name.mjs` argument to Node.
+provided via a `--loader ./loader-name.mjs` argument to Node. This argument
+can be passed multiple times to compose loaders like
+`--loader ./loader-coverage.mjs --loader ./loader-mocking.mjs`. The last loader
+must explicitly call to the parent loader in order to provide compose behavior.
```

Comment:
I actually like the current design - it pretty much mimics the behavior of middleware in `koa`, which is kinda similar.

Reply:
@devsnek `== null`, probably :-)

@weswigham the difference I see here is that if a middleware breaks the contract, your entire request likely fails, because nothing further proceeds; however, depending on what the loaders do, node might silently do the wrong thing.

Reply:
We should definitely not just check for undefined. We should require full responsibility to handle all cases and avoid polymorphic return types. We could add something more complicated to model having defaults, but I'm more curious about a code example of this idea of the current design being a pit of failure.

Since the parameters are not sufficient to produce a valid return value, loaders must do *something* to return a proper value. Is there a real world use case where you are not breaking the chain but are also not fundamentally always calling the parent to do useful things?

Reply:
@ljharb I don't think anyone can do "the wrong thing" - a loader can simply choose to ignore other loaders and produce a result without consulting them. Now, you could wrap the handlers passed into the resolver as the parent handler and have the wrapped version record that it was called, and then check if it was called after the hook returns to require that it gets called (even if the result is ignored), but I'm not sure that's needed.

Reply:
@weswigham I'd rather investigate something along the lines of `event.preventDefault()` rather than trying to figure out polymorphic return types and/or magical default behavior around if the parent was called. Unfortunately those APIs would probably make the parameters passed to hooks more complex.

Reply:
`event.preventDefault()` is a pretty awful API and doesn't let you do things both before _and_ after the other pipelined event executes.

Reply:
fwiw I agree that all of the use cases where "being able to explicitly call the parent loader" are valuable, including "not calling the parent loader" - however, I think it's important that the default, when no explicit choice is made, should be "call the parent loader".

Reply:
@ljharb the isn't really a way to explicitly say you are not choosing your return value. We can only make well known return values, which I would want to be the same type as the other return values.

Reply:
@ljharb to clarify. I am not really seeing the case where you want to default to the parent, but do not want to take any action on it as the common case. Most loaders want to redirect behaviors or mutate what would be loaded. The only way to know what would be loaded is to query the parent.

Reply:
> however, I think it's important that the default, when no explicit choice is made, should be "call the parent loader".

I'm not sure that's correct - If my loader wasn't made with a loader pipeline in mind, how am I even sure the loader is capable of yielding output another loader can usefully use?

Reply:
Like `return null` is probably more cryptic than `return parentResolve()`.